src/redis/windowed_cache_impl.go

```diff
@@ -0,0 +1,168 @@
+	perSecondClient            driver.Client
+	timeSource                 utils.TimeSource
+	jitterRand                 *rand.Rand
+	expirationJitterMaxSeconds int64
+	cacheKeyGenerator          utils.CacheKeyGenerator
+	localCache                 *freecache.Cache
+	nearLimitRatio             float32
+	algorithm                  algorithm.RatelimitAlgorithm
+}
+
+func (this *windowedRateLimitCacheImpl) DoLimit(
+	ctx context.Context,
+	request *pb.RateLimitRequest,
+	limits []*config.RateLimit) []*pb.RateLimitResponse_DescriptorStatus {
+
+	logger.Debugf("starting windowed cache lookup")
+
+	// request.HitsAddend could be 0 (default value) if not specified by the caller in the Ratelimit request.
+	hitsAddend := utils.MaxInt64(1, int64(request.HitsAddend))
+
+	// First build a list of all cache keys that we are actually going to hit.
+	cacheKeys := this.algorithm.GenerateCacheKeys(request, limits, hitsAddend)
+
+	isOverLimitWithLocalCache := make([]bool, len(request.Descriptors))
+	tats := make([]int64, len(request.Descriptors))
+	var pipeline, perSecondPipeline driver.Pipeline
+
+	// Now, actually setup the pipeline, skipping empty cache keys.
+	for i, cacheKey := range cacheKeys {
+		if cacheKey.Key == "" {
+			continue
+		}
+
+		// Check if key is over the limit in local cache.
+		if this.algorithm.IsOverLimitWithLocalCache(cacheKey.Key) {
+			isOverLimitWithLocalCache[i] = true
+			logger.Debugf("cache key is over the limit: %s", cacheKey.Key)
+			continue
+		}
+
+		logger.Debugf("looking up tat for cache key: %s", cacheKey.Key)
+
+		expirationSeconds := utils.UnitToDivider(limits[i].Limit.Unit)
+
+		// Use the perSecondConn if it is not nil and the cacheKey represents a per second Limit.
+		if this.perSecondClient != nil && cacheKey.PerSecond {
+			if perSecondPipeline == nil {
+				perSecondPipeline = driver.Pipeline{}
+			}
+			windowedPipelineAppend(this.perSecondClient, &perSecondPipeline, cacheKey.Key, &tats[i], expirationSeconds)
+		} else {
+			if pipeline == nil {
+				pipeline = driver.Pipeline{}
+			}
+			windowedPipelineAppend(this.client, &pipeline, cacheKey.Key, &tats[i], expirationSeconds)
+		}
+	}
+
+	if pipeline != nil {
+		driver.CheckError(this.client.PipeDo(pipeline))
+		pipeline = nil
+	}
+	if perSecondPipeline != nil {
+		driver.CheckError(this.perSecondClient.PipeDo(perSecondPipeline))
+		perSecondPipeline = nil
+	}
+
+	responseDescriptorStatuses := make([]*pb.RateLimitResponse_DescriptorStatus, len(request.Descriptors))
+
+	for i, cacheKey := range cacheKeys {
+		responseDescriptorStatuses[i] = this.algorithm.GetResponseDescriptorStatus(cacheKey.Key, limits[i], int64(tats[i]), isOverLimitWithLocalCache[i], int64(hitsAddend))
+
+		if cacheKey.Key == "" || isOverLimitWithLocalCache[i] {
+			continue
+		}
+
+		arrivedAt := this.algorithm.GetArrivedAt()
+		newTat := this.algorithm.GetNewTat()
+
+		// Store new tat for initial tat of next requests
+		expirationSeconds := utils.NanosecondsToSeconds(newTat-arrivedAt) + 1
```

Comment:
you could make algorithm object responsible for calculating expirationSeconds, that way you don't need to expose `GetArrivedAt()`, `GetNewTat()` as part of interface.

Reply:
This is not the cases, we need `newTat` to be added to the redis/memcached after all.

Reply:
I am removing GetArrivedAt() & GetNewTat() and creating GetExpirationSeconds() & GetResultsAfterIncrease()